control-plane-operator/hostedclusterconfigoperator/controllers/machine/machine.go

```diff
@@ -0,0 +1,174 @@
+		hcp.Spec.Platform.Kubevirt.GenerateID)
+
+	if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(cpService), cpService); err != nil {
+		return fmt.Errorf("failed to get default ingress passthrough Service: %w", err)
+	}
+
+	// If there is a selector endpoints should not be generated
+	if len(cpService.Spec.Selector) > 0 {
+		return nil
+	}
+
+	if len(cpService.Spec.Ports) == 0 {
+		return fmt.Errorf("missing default ingress passthrough Service %s/%s ports", cpService.Namespace, cpService.Name)
+	}
+
+	ipv4MachineAddresess := []string{}
+	ipv6MachineAddresses := []string{}
+	ports := []discoveryv1.EndpointPort{}
+
+	// If machine is back to not ready we should reset the endpointslice
+	if machine.Status.GetTypedPhase() == capiv1.MachinePhaseRunning {
+		for _, machineAddress := range machine.Status.Addresses {
+			if machineAddress.Type == capiv1.MachineInternalIP {
+				if netip.MustParseAddr(machineAddress.Address).Is4() {
+					ipv4MachineAddresess = append(ipv4MachineAddresess, machineAddress.Address)
+				} else {
+					ipv6MachineAddresses = append(ipv6MachineAddresses, machineAddress.Address)
+				}
+			}
+		}
+		for _, port := range cpService.Spec.Ports {
+			ports = append(ports, discoveryv1.EndpointPort{
+				Protocol: &port.Protocol,
+				Port:     pointer.Int32(int32(port.TargetPort.IntValue())),
+			})
+		}
+	}
+
+	if serviceHasIPFamily(cpService, corev1.IPv4Protocol) {
+		if err := r.reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx, machine, cpService, corev1.IPv4Protocol, ipv4MachineAddresess, ports); err != nil {
+			return err
+		}
+	}
+
+	if serviceHasIPFamily(cpService, corev1.IPv6Protocol) {
+		if err := r.reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx, machine, cpService, corev1.IPv6Protocol, ipv6MachineAddresses, ports); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcileKubevirtDefaultIngressEndpointsByIPFamily(ctx context.Context, machine *capiv1.Machine, cpService *corev1.Service, ipFamily corev1.IPFamily, machineAddresses []string, ports []discoveryv1.EndpointPort) error {
+	log := ctrl.LoggerFrom(ctx)
+	endpointSlice := &discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cpService.Namespace,
+			Name:      cpService.Name + "-" + machine.Name + "-" + strings.ToLower(string(ipFamily)),
+		},
+	}
+	result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, endpointSlice, func() error {
+		if len(endpointSlice.OwnerReferences) == 0 {
+			vmKey := client.ObjectKey{
+				Namespace: machine.Spec.InfrastructureRef.Namespace,
+				Name:      machine.Spec.InfrastructureRef.Name,
+			}
+			vm := kubevirtv1.VirtualMachine{}
+			if err := r.kubevirtInfraClient.Get(ctx, vmKey, &vm); err != nil {
+				//TODO: Should we skip NotFound ? At the end Machine is going
+				//      to be reconcile later on.
+				return fmt.Errorf("failed retrieving kubevirt virtual machine to compose default ingress passthrough endpointslices owner ref: %w", err)
```

Comment:
Yeah, i think we might need to account for the brief gap in time where the Machine exists but the VM hasn't been created yet. As this is written now, I believe it will always result in some reconcile errors when a Machine is first created due to that gap.

Maybe this is as simple as returning the error here, which should prevent the CreateOrUpdate function from creating the endpointslice. 

```
            if err := r.kubevirtInfraClient.Get(ctx, vmKey, &vm); err != nil {
				return err
            }
```

Then when interpreting the error from CreateOrUpdate, only return an error from `reconcileKubevirtDefaultIngressEndpointsByIPFamily` when the error something other than not found.
```
        if apierrors.IsNotFound(err) {
                log.Info(fmt.Sprintf("waiting for kubevirt VM to be created before processing default ingress %s endpoints", ipFamily), "result", result)
        } else if err != nil {
               return fmt.Errorf("failed to reconcile kubevirt default ingress %s endpoints: %w", ipFamily, err)
        }

```